astropy/units/utils.py

```diff
@@ -6,21 +6,37 @@
 package.
 """
 
+from __future__ import annotations
+
 import io
 import re
 from fractions import Fraction
+from typing import TYPE_CHECKING, overload
 
 import numpy as np
 from numpy import finfo
 
+if TYPE_CHECKING:
+    from collections.abc import Generator, Sequence
+    from numbers import Complex, Real
```

Comment:
I recently learned that [PEP 484](https://peps.python.org/pep-0484/#the-numeric-tower) recommends using `float` instead of `numbers.Real` and `complex` instead of `numbers.Complex` for type hints 🤔.  There's an older discussion about the reasons why in https://github.com/dropbox/stone/issues/33, though I believe `mypy` now works satisfactorily in both cases.

Reply:
The last sentence of the paragraph you linked to is:

>  This does not handle classes implementing the corresponding ABCs or the fractions.Fraction class, but we believe those use cases are exceedingly rare.

But the functions here do need to handle `Fraction`.

Reply:
The python numeric tower is a mess! But this looks correct.